lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp

```diff
@@ -411,10 +411,60 @@ struct VerbatimBuilder {
   struct Base {
     SmallString<128> string;
     SmallVector<Attribute> symbols;
+    SmallVector<Value, 2> subst;
     VerbatimBuilder builder() { return VerbatimBuilder(*this); }
     operator VerbatimBuilder() { return builder(); }
   };
 
+  /// Increment all the indices inside `{{`, `}}` by one. This is to indicate
+  /// that a value is added to the `substitutions` of the verbatim op, other
+  /// than the symbols.
+  void incrementIds() {
+    StringRef begin = "{{";
+    StringRef end = "}}";
+    // The replacement string.
+    size_t from = 0;
+    replStr.clear();
+    while (from < base.string.size()) {
+      // Search for the first `{{` and `}}`.
+      auto beginAt = base.string.find(begin, from);
+      auto endAt = base.string.find(end, beginAt);
+      // Copy the string as is, until the `{{`.
+      replStr.append(base.string.substr(from, beginAt - from));
+      // If not found, then done.
+      if (beginAt == StringRef::npos || endAt == StringRef::npos)
+        break;
+      // Advance `from` to the character after the `}}`.
+      from = endAt + 2;
+      auto idChar = base.string.substr(beginAt + 2, endAt - beginAt - 2);
+      unsigned idNum;
+      bool failed = idChar.getAsInteger(10, idNum);
+      assert(!failed && "failed to parse integer from verbatim string");
+      // Now increment the id and append.
+      replStr.append("{{");
+      Twine(idNum + 1).toVector(replStr);
+      replStr.append("}}");
+    }
+  }
+
+  /// Add Value for substitution. Since the value precedes the symbols in the
+  /// index, increment all the ids in the string, and then append the value for
+  /// index 0. The original base.string is not updated, but a copy replStr is
+  /// maintained for value sustitution. The original base.string is not updated,
+  /// since the snapshot should still return the original un-modified string.
+  VerbatimBuilder &addValue(Value s) {
+    // Add the value to be used for substitution.
+    base.subst.push_back(s);
+    // Create a copy of the base.string into replStr, and increment all the
+    // index integers by 1.
+    incrementIds();
```

Comment:
I think there's a problem if anyone tries to use `append` after any substitution values are added?

1) `replString` will be out-of-sync
2) Placeholder strings (see `append(Attribute symbol)`) will be generated using `base.symbols.size()` when I guess it should be `base.subst.size() + base.symbols.size()` if keeping this approach.